homeassistant/components/gree/__init__.py

```diff
@@ -21,52 +31,61 @@ async def async_setup(hass: HomeAssistant, config: dict):
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Set up Gree Climate from a config entry."""
-    devices = []
+    gree_discovery = DiscoveryService(hass)
+    hass.data[DATA_DISCOVERY_SERVICE] = gree_discovery
 
-    # First we'll grab as many devices as we can find on the network
-    # it's necessary to bind static devices anyway
-    _LOGGER.debug("Scanning network for Gree devices")
+    @callback
+    def shutdown_event(_: Event) -> None:
+        if hass.data[DOMAIN].get(DATA_DISCOVERY_INTERVAL) is not None:
+            hass.data[DOMAIN][DATA_DISCOVERY_INTERVAL]()
+            hass.data[DOMAIN].pop(DATA_DISCOVERY_INTERVAL)
+
+        if hass.data.get(DATA_DISCOVERY_SERVICE) is not None:
+            del hass.data[DATA_DISCOVERY_SERVICE]
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown_event)
```

Comment:
Remove the subscription in `async_unload_entry`.

Something like this: https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/__init__.py#L394-L395